scala/src/main/org/apache/spark/streaming/api/kafka/ParallelKafkaDataFetcherDStream.scala

```diff
@@ -0,0 +1,270 @@
+        case (tp, fo) => fetchResult.untilOffsets(tp).offset - fo
+      }.sum
+      rddList += record.rdd
+    }
+    logInfo(s"Get ${rddList.length} pre-computed RDDs for DStream-$id," +
+      s" totalMessageCount: $totalMessageCount")
+
+    // remember the precomputed output for later un-persist and ack operation
+    preComputeOutputMap.put(validTime, preComputeOutput)
+
+    val rdd: RDD[R] = {
+      if (rddList.nonEmpty) {
+        context.sparkContext.union(rddList)
+      } else {
+        context.sparkContext.emptyRDD[R]
+      }
+    }
+    (rdd, totalMessageCount)
+  }
+
+  private def initialize(): Unit = {
+    if (!initialized) {
+      if (OffsetsFetcher == null) {
+        OffsetsFetcher = new KafkaOffsetsFetcher(
+          context, fetcherStateSnapshot, refreshOffsetsInterval, kafkaParams, topics, fromOffsets)
+        OffsetsFetcher.start()
+      }
+
+      if (precomputeCache == null) {
+        precomputeCache = new RddPreComputeCache[R](context.sparkContext,
+          s"ParallelKafkaDataFetcher-$id", maxUnackedNum, spitLimit, numThreads,
+          StorageLevel(useDisk = false, useMemory = true, deserialized = true, 3))
+        precomputeCache.start()
+      }
+
+      if (offsetsScheduler == null) {
+        offsetsScheduler = ThreadUtils.newDaemonSingleThreadScheduledExecutor(
+          "offsetsScheduler-" + UUID.randomUUID().toString.substring(0, 8))
+        offsetsScheduler.execute(new Runnable {
+          override def run(): Unit = {
+            while (true) {
+              val fetchResult = OffsetsFetcher.take()
+              preCompute(fetchResult)
+            }
+          }
+        })
+      }
+
+      initialized = true
+    }
+  }
+
+  override def compute(validTime: Time): Option[RDD[R]] = {
+    val preComputeOutput = precomputeCache.spit()
+    val (rdd, totalMessageCount) = processRddPreComputeOutput(validTime, preComputeOutput)
+    val inputInfo = StreamInputInfo(id, totalMessageCount)
+    ssc.scheduler.inputInfoTracker.reportInfo(validTime, inputInfo)
+    Some(rdd)
+  }
+
+  // override this method to un-persist precomputed RDDs that are no longer needed
+  override private[streaming] def clearMetadata(time: Time): Unit = {
+    if (preComputeOutputMap.containsKey(time)) {
+      for (record <- preComputeOutputMap.get(time)) {
+        record.rdd.unpersist(false)
+        precomputeCache.ackSeqNum(record.rddSeqNum)
+      }
+    }
+    super.clearMetadata(time)
+  }
+
+  override def start(): Unit = {
+    initialize()
+  }
+
+  override def stop(): Unit = {
+    if (OffsetsFetcher != null) {
+      OffsetsFetcher.stop()
+      OffsetsFetcher = null
+    }
+
+    if (precomputeCache != null) {
+      precomputeCache.stop()
+      precomputeCache = null
+    }
+
+    if (offsetsScheduler != null) {
+      offsetsScheduler.shutdown()
+      offsetsScheduler = null
+    }
+  }
+
+  private[streaming]
+  class ParallelKafkaDataFetcherCheckpointData extends DStreamCheckpointData(this) {
```

Comment:
Great to see that you are extending checkpoint data and making sure the implementation of the new DStream is robust